brainiak/utils/fmrisim.py

```diff
@@ -2134,126 +2701,229 @@ def generate_noise(dimensions,
-        supplied
 
     Returns
     ----------
-    ax : matplotlib object
-        Object with the information to be plotted
+    signal_function_scaled : 4d numpy array
+        The new signal volume with the appropriately set signal change
 
     """
 
-    ax = fig.add_subplot(111, projection='3d')
+    # If you have only one magnitude value, duplicate the magnitude for each
+    #  timecourse you have
+    if len(magnitude) == 1:
+        magnitude *= signal_function.shape[1]
 
-    # Threshold the data
-    threshold = np.percentile(brain.reshape(np.prod(brain.shape[0:3])),
-                              percentile)
+    # Scale all signals that to have a range of -1 to 1. This is
+    # so that any values less than this will be scaled appropriately
+    signal_function /= np.max(np.abs(signal_function))
 
-    # How many voxels exceed a threshold
-    brain_threshold = np.where(np.abs(brain) > threshold)
+    # Iterate through the timecourses and calculate the metric
+    signal_function_scaled = np.zeros(signal_function.shape)
+    for voxel_counter in range(signal_function.shape[1]):
 
-    # Clear the way
-    ax.clear()
+        # Pull out the values for this voxel
+        sig_voxel = signal_function[:, voxel_counter]
+        noise_voxel = noise_function[:, voxel_counter]
+        magnitude_voxel = magnitude[voxel_counter]
 
-    ax.set_xlim(0, brain.shape[0])
-    ax.set_ylim(0, brain.shape[1])
-    ax.set_zlim(0, brain.shape[2])
+        # Calculate the scaled time course using the specified method
+        if method == 'SFNR':
 
-    # If a mask is provided then plot this
-    if mask is not None:
-        mask_threshold = np.where(np.abs(mask) > 0)
-        ax.scatter(mask_threshold[0],
-                   mask_threshold[1],
-                   mask_threshold[2],
-                   zdir='z',
-                   c='black',
-                   s=10,
-                   alpha=0.01)
-
-    # Plot the volume
-    ax.scatter(brain_threshold[0],
-               brain_threshold[1],
-               brain_threshold[2],
-               zdir='z',
-               c='red',
-               s=20)
-
-    return ax
+            # How much temporal variation is there, relative to the mean
+            # activity
+            temporal_var = noise_voxel.mean() / noise_dict['sfnr']
+
+            # Multiply the timecourse by the variability metric
+            new_sig = sig_voxel * (temporal_var * magnitude_voxel)
+
+        elif method == 'CNR_Amp/Noise-SD':
+
+            # What is the standard deviation of the noise
+            noise_std = np.std(noise_voxel)
+
+            # Multiply the signal timecourse by the the CNR and noise (
+            # rearranging eq.)
+            new_sig = sig_voxel * (magnitude_voxel * noise_std)
+
+        elif method == 'CNR_Amp2/Noise-Var_dB':
+
+            # Calculate the current signal amplitude (likely to be 1,
+            # but not necessarily)
+            sig_amp = np.max(np.abs(sig_voxel))
```

Comment:
CNR_Amp/Noise-SD seems to calculate CNR based on the change of signal in the voxel with the largest change (becuase every voxel is scaled only to the maximum absolute value of the whole signal_function). But here CNR seems to be calculated for each voxel separately because sig_amp is calculated for each voxel. I am not sure which is better. The current approach can have problem if signal_function has all-zero time course in any voxel. The approach in CNR_Amp/Noise-SD can also be a bit wierd if signal_function have different magnitude in different voxels originally and if the magnitude parameter is a vector -- the final CNR for each voxel won't be what is asked for in the magnitude parameter. But regardless of which one you choose, I think it may be better that different metrics take consistent way of scaling, and it may be good to specify what is required for signal_function (is it expected that they only reflect the hypothetical time courses of signals in each voxel but they should not reflect magnitude difference across voxels?)

Reply:
Wow, great catch, this is perceptive. I decided to scale the CNR choices to the maximum amplitude (which is almost always 1 anyway). I added a note to the description to clarify that voxels values are all relative to each other